Update DiagnosticPipError to add details_link to errors (#14002)
pypa/pip: docs/html/errors.md

# Error Index

Inspired by the [Rust Compiler Error Index], this page describes the various

> errors that may be presented by `sphinx-theme-builder`, indicating known causes

Reviewer: ```suggestion
errors that may be presented by `pip`, indicating known causes
```

as well as potential solutions.

<!--
  Editor's note: Ensure that each error has a "What you can do" part to it.
-->

[rust compiler error index]: https://doc.rust-lang.org/error-index.html

## Pip Errors

Generic pip errors do not contain links to this page.

## Diagnostic Pip Errors

> Diagnositic pip errors contain links to this page.

Reviewer: We can drop these. 😅

### externally-managed-environment

An `externally-managed-environment` error occurs when your operating system (OS) prevents you from using pip to install packages directly into the system-wide Python environment. This restriction was introduced via PEP 668 to prevent users from accidentally breaking OS-level tools that rely on a specific version of Python.

**What you can do:**

> There are three standard ways to resolve this, depending on your goal: First, use a Virtual Environment. Second, use pipx if you want to install a tool for use from the command line. Lastly, if you absolutely must install a package system-wide and understand the risks of breaking your OS, then you can bypass the check by adding the specific flag: `--break-system-packages` to the pip install command.

Reviewer: Let's make this a bullet list.

### failed-build-dependency-install

A `failed-build-dependency-install` error typically means pip cannot find or compile the libraries necessary to complete the installation of the dependency.

**What you can do:**

Check that the python version is compatible with the version of the python dependency you are attempting to install. Upgrading pip with the `pip install --upgrade pip` command will often resolve failed builds. Check the documentation of the dependency for whether C/C++ compilers are needed to properly install the dependency.

### failed-wheel-build-for-install

When pip cannot find a wheel file for a dependency, pip tries to build the wheel from the source. If building the wheel from source fails, this error is raised.

> **What you can do:**

Reviewer: nit: Add newline after, for consistency -- here, and elsewhere.

### invalid-egg-fragment

The invalid-egg-fragment error occurs because modern versions of pip (specifically version 25.0 and later) have removed support for complex #egg= fragments in URLs. This fragment was previously used to specify package names and extras in VCS (Version Control System) links, but it has been deprecated in favor of more modern standards.

**What you can do:**

Replace the old fragment-style URL with the PackageName @ URL format:

Old (Broken): `pip install git+https://github.com[extra]`

New (Correct): `pip install "pkgname[extra] @ git+https://github.com"`

If a subdirectory is used, the full format becomes `pip install "pkgname[extra] @ git+https://github.com/user/project.git#subdirectory=path/to/subdir"`

### invalid-installed-package

The "invalid-installed-package" error (often appearing as an InvalidRequirement or InvalidDistribution warning) typically occurs when pip encounters a corrupted or improperly named folder in your Python site-packages directory. This often happens due to failed uninstalls or interrupted upgrades.

**What you can do:**

> Clear the pip cache `pip cache purge` and then upgrade pip `pip install --upgrade pip`. Then find the path to the site-packages `python -m site` and look for folders with invalid names such as folders with `~` and delete them.

Reviewer: ```suggestion
Clear the pip cache (`pip cache purge`) and then upgrade pip (`pip install --upgrade pip`). Then find the path to the site-packages `python -m site` and delete folders with invalid names such as folders with `~`.
```

### invalid-pyproject-build-system-requires

The error invalid-pyproject-build-system-requires occurs when pip encounters a `pyproject.toml` file where the `[build-system]` section has an incorrectly formatted `requires` key.

**What you can do:**

Ensure requires is a list: Check your pyproject.toml and verify that requires is wrapped in brackets. Specify the build-backend such as `build-backend = "setuptools.build_meta"` when applicable.

> Incorrect: requires = "setuptools"
> Correct: requires = ["setuptools", "wheel"]
> 

Reviewer: ```suggestion

Incorrect: `requires = "setuptools"`
Correct: `requires = ["setuptools", "wheel"]`
```

### metadata-generation-failed

The `metadata-generation-failed` typically occurs when when pip is unable to generate or read the metadata. This error can happen when the metadata files are corrupted or missing, when the package has an incorrect structure, or a non python dependency is not present.

**What you can do:**

Check that the python version is compatible with the dependencies that you are attempting to install. Ensure the build tools are up to date by running `pip install --upgrade pip setuptools wheel`.

### missing-pyproject-build-system-requires

The error "missing-pyproject-build-system-requires" (or warnings related to it) typically occurs when your pyproject.toml file is missing the mandatory [build-system] table, which pip needs to identify how to build and install your package

**What you can do:**

Add the following block or equivalent to your pyproject.toml file. This tells pip to use setuptools as the default build tool:

> ```

Reviewer: ````suggestion
```toml
````
